knox.go

```diff
@@ -145,6 +166,79 @@ func (s PrincipalType) MarshalJSON() ([]byte, error) {
 	}
 }
 
+// IsValidPrincipal verifies that the given id string matches our expectations
+// for what a principal should look like given the principal type. For example,
+// a service principal should be a valid SPIFFE ID.
+func (s PrincipalType) IsValidPrincipal(id string, extraValidators []PrincipalValidator) error {
+	// Empty principals are invalid. Note we must deny addition of ACLs with
+	// empty MachinePrefix to prevent abuse. An empty prefix would grant *every*
+	// client access, which we don't support in Knox (if it's not a secret it
+	// doesn't belong in Knox). For other types of principals, an empty string is
+	// also invalid (though would not broadly open up access).
+	if id == "" {
+		return ErrACLEmptyPrincipal
+	}
+
+	// Apply additional validation for service and service prefix principals.
+	switch s {
+	case Service, ServicePrefix:
+		// For Service principals and prefixes, verify that id looks like a valid SPIFFE ID.
+		parsed, err := url.Parse(id)
+		if err != nil || parsed.Scheme != spiffeScheme || parsed.Host == "" {
+			return ErrACLInvalidServicePrefixURL
+		}
```

Comment:
I think you could scope this variables to the condition?

```go
if parsed, err := url.Parse(id); err != nil || parsed.Scheme != spiffeScheme || parsed.Host == "" {
```

Reply:
I could, but I felt it was more readable to break it out.

Reply:
that's a long / ugly line imo, I think the existing is fine. 🤷